### ISSUE TITLE
[FEATURE] Make the @SuperBuilder's *BuilderImpl class public

### ISSUE THREAD

**Issue by ilia1243** [projectlombok/lombok#3791]
**Describe the feature**

@SuperBuilder's *BuilderImpl class is currently private, and cannot be used as a generic type argument. Making it public will allow to use it in the following way:

```java
@SuperBuilder
abstract class Base {
}

@SuperBuilder
class Derived extends Base {
}

abstract class BaseService<T extends Base, B extends Base.BaseBuilder<T, B>> {
    abstract B newBuilder();
}

class DerivedService extends BaseService<Derived, Derived.DerivedBuilderImpl> {
    @Override
    public Derived.DerivedBuilderImpl newBuilder() {
        return Derived.builder();
    }
}
```

**Note**: this will also allow for `.builder()` to return `*BuilderImpl` class instead of `*Builder<?, ?>`. This will likely fix issue #3445.

**Workaround**.

The above code will compile if the derived class has customized `.builder()` method, and customized `*BuilderImpl` class.

```java
@SuperBuilder
class Derived extends Base {
    public static DerivedBuilderImpl builder() {
        return new DerivedBuilderImpl();
    }

    public static class DerivedBuilderImpl extends DerivedBuilder<Derived, DerivedBuilderImpl> {
    }
}
```

**Comment by janrieke** [comment]
Yeah, well, but where is the use case? I don't understand why you'd want to do this.
However, to be fair, probably nothing will convince me that this is a good idea. Making that `BuilderImpl` visible for the world will create _a lot_ of confusion. But if you'd describe your use case, we may figure out another way that doesn't lead to users having a thousand ways of messing up a `@SuperBuilder`.

**Comment by ilia1243** [comment]
I was not accurate enough. In the issue description, it is possible to use abstract builder class with wildcards.

```java
abstract class BaseService<B extends Base.BaseBuilder<?, ?>> {
    abstract B newBuilder();
}

class DerivedService extends BaseService<Derived.DerivedBuilder<?, ?>> {
    public Derived.DerivedBuilder<?, ?> newBuilder() {
        return Derived.builder();
    }
}
```

But `BaseService.newBuilder()` cannot be used inside the `BaseService` itself. Consider a bit more complex example:

```java
@SuperBuilder
public abstract class Base {
    private boolean base;
}

@SuperBuilder
public class Derived extends Base {
}

public abstract class BaseService<B extends Base.BaseBuilder<?, ?>> {
    protected abstract B createBuilder();

    public B newBuilder() {
        return createBuilder()
                .base(true);
    }
}

public class DerivedService extends BaseService<Derived.DerivedBuilder<?, ?>> {
    @Override
    protected Derived.DerivedBuilder<?, ?> createBuilder() {
        return Derived.builder();
    }
}
```

This does not complie with error in BaseService class:
```
BaseService.java:8: error: incompatible types: CAP#1 cannot be converted to B
                .base(true);
                     ^
```

If using recurring generics (this fixes the compilation error in BaseService) `class BaseService<B extends Base.BaseBuilder<?, B>>` I came to the problem in the issue description. Abstract builder class cannot be used as a type argument in the DerivedService. For example, `class DerivedService extends BaseService<Derived.DerivedBuilder<?, ?>>` produces:
```
DerivedService.java:3: error: type argument DerivedBuilder<?,?> is not within bounds of type-variable B
public class DerivedService extends BaseService<Derived.DerivedBuilder<?, ?>> {
                                                                      ^
```

I currently see only one way to fix this by making DerivedBuilderImpl public and using as type argument `class DerivedService extends BaseService<Derived.DerivedBuilderImpl>`, though probably missing something.

**Comment by janrieke** [comment]
The issue with using `DerivedBuilderImpl` as type argument at that point is that you cannot have another "sub-service", e.g. `SubDerivedService extends DerivedService`. It does not work because the `SubDerivedBuilderImpl` is not a sub-class of `DerivedBuilderImpl`, but only of `DerivedBuilder`.

**Comment by ilia1243** [comment]
Here is the way to have deeper inheritance (Base and BaseService are omitted for brevity).

```java
@SuperBuilder(builderMethodName = "derivedBuilder")
public class Derived extends Base {
}

@SuperBuilder(builderMethodName = "subDerivedBuilder")
public class SubDerived extends Derived {
}

public abstract class AbstractDerivedService<B extends Derived.DerivedBuilder<?, B>> extends BaseService<B> {
}

public class DerivedService extends AbstractDerivedService<Derived.DerivedBuilderImpl> {
    @Override
    protected Derived.DerivedBuilderImpl createBuilder() {
        return Derived.derivedBuilder();
    }
}

public class SubDerivedService extends AbstractDerivedService<SubDerived.SubDerivedBuilderImpl> {
    @Override
    protected SubDerived.SubDerivedBuilderImpl createBuilder() {
        return SubDerived.subDerivedBuilder();
    }
}
```
* This still requires to have BuilderImpl public in both Derived and SubDerived (I omitted this). `builder()` method also returns `BuilderImpl` in both classes.
* I had to use different `builderMethodName` in Derived, and SubDerived. Otherwise, I cannot return `BuilderImpl` due to compule error `error: builder() in SubDerived cannot hide builder() in Derived ... return type SubDerivedBuilderImpl is not compatible with DerivedBuilderImpl`.

In fact, it seems unnatural, as we do not really have instances of Animals, we have instances of Cats that we can work with as Animals. I mean, in practice Derived will likely also be abstract, and there will be no final DerivedService in this case. This way there will be no need to have different `builderMethodName`.

**Comment by janrieke** [comment]
Sorry for the late reply, I missed your answer.

Your last proposals still add clutter and confusion for the average lombok user. Adding a second builder method is definitely a no-go. Making the constructor public would mean having two ways for exactly the same thing, and that's also not going to happen: Lombok should always promote a good coding style, and that proposal is an anti-pattern.

I can understand your use case, but given that there is a simply workaround (customizing the builder of the subclass), the disadvantages clearly outweigh potential advantages.

**Comment by ilia1243** [comment]
What about only making the `BuilderImpl` class public? There will be only one way to instantiate the builder, and the class can still be used as a type argument.

```java
public class DerivedService extends BaseService<Derived, Derived.DerivedBuilderImpl> {
    @Override
    public Derived.DerivedBuilderImpl newBuilder() {
        return (Derived.DerivedBuilderImpl) Derived.builder();
    }
}
```

**Comment by janrieke** [comment]
That would also end up in the auto-completion dialog of _all_ users. The only thing I can imagine is some signal in the source code that will make Lombok recognize the intention of the user to have a public builder impl. That's the only way not to have 99 % of the users getting unnecessary and confusing auto completions.
However, that would be kind of a novel approach for Lombok, so I don't know what the maintainers think about that.

**Comment by ilia1243** [comment]
Than you for the detailed answer, closing the issue.
